docs: reorganize documentation, add new guides (#17)
threeport/user-docs: docs/index.md

# Threeport

Threeport is an Application Orchestrator.

App Orchestration is the next evolution for software delivery.  For years we have used
continuous delivery pipelines and GitOps as methods for delivering software.  As
cloud native has become more ubiquitous, these methods have become increasingly
insufficient.

Threeport offers a self-service experience to developers.  The developer's

> software delivery is facilitated by DevOps and platform engineering teams, but

Reviewer: We use both "DevOps" and "Platform Engineering" in different ways, but I don't think we define how exactly we see the two terms relating to one another. Based on what we have written, it sounds like platform engineering would be those writing code to support software delivery, and devops is more of an admin role that defines object definitions for use by platform & development teams. Does that sound right?

Author: Yep.  That's right.  Delving into terms like this seems like a bit of a sidetrack in the Threeport docs.  I think that kind of content would be best served in a blog post (and these topics are planned).  When those posts are up, we'll add links to the docs to reference that stuff.

does _not_ put those teams on the day-to-day critical delivery path for

> applications.

Reviewer: ```suggestion
Threeport offers a self-service experience to developers.  Software delivery can
still be administered by DevOps (or platform engineering) teams, but does _not_
put those teams on the day-to-day critical delivery path for applications.
```

Just trying to emphasize how work _could_ be delegated to other teams, but the day-to-day delivery path is independent regardless of how that is set up.

Author: Good point.  I re-worded this passage a bit more to convey this.

   Threeport will call the Kubernetes API to deploy the Kubernetes resources.
   If requested, Threeport can deploy AWS managed service dependencies as well
   as other support services to configure ingress routing from the internet,
   provision SSL certs, set DNS records, etc.
0. The image is pulled by the Kubernetes control plane.
0. The app runs in containers on the Kubernetes cluster.  This example shows
   the app running in the same Kubernetes environment as the Threeport Control
   Plane, but this is optional.  Threeport can deploy separate Kubernetes
   runtime environments as needed, and manage deployments to any cluster managed
   by Threeport.

## Threeport for DevOps

DevOps can streamline developer usage by creating definitions for the resources
developers need to deliver.  Most Threeport objects have two components: a
definition and instance.  The definition provides the configuration.  The
instance uses a small number of runtime parameters and references the definition
to spin up the resources required.  Learn more about definitions and instances
in the [Concepts section](concepts/definitions-instances.md).

![Threeport for DevOps](img/ThreeportForDevOps.png)

> Following is an example of this works:

Reviewer: ```suggestion
The following is an example of this works:
```

Author: Modified this sentence to be more relevant in separate commit.

0. DevOps creates the definition configurations that can be referenced later by
   developers.  Any number of definitions can be made available for different
   needs for each resource type.
0. The developer creates instances from these definitions to create a new
   Kubernetes runtime, if needed, service dependencies like network gateways,
   managed dependencies like RDS databases, as well as the workload itself.  The
   developer doesn't have to worry about any resource configuration since DevOps
   took care of this.  They only provide runtime parameters such as which
   Kubernetes runtime to use for their workload.
0. Threeport deploys the Kubernetes runtime and, if needed, the RDS database, the
   gateway controller along with its load balancer and the workload itself.  All
   dependencies are connected so that the workload is immediately available to

>    the end user as soon as the resources are up.

Reviewer: ```suggestion
0. Threeport deploys a kubernetes runtime and any dependencies that may be
   needed. These dependencies can include kubernetes manifests, gateway
   resources such as load balancers, and managed AWS services such as RDS.  All
   dependencies are connected and the workload is immediately available to the
   end user as soon as resources are up. An example configuration can be found
   on the Threeport GitHub repository
   [here](https://github.com/threeport/threeport/blob/main/samples/wordpress-workload-remote.yaml).
```

The process for optimizing application delivery with platform engineering looks
something like this:

0. A platform engineering team uses the Threeport SDK to build an extension to the
   Threeport control plane.  The code for this extension lives in its own
   git repo.  Experienced Go programmers with a sound understanding of
   Kubernetes and Threeport can usually produce a POC within a couple of weeks,
   even faster for simpler use cases.
0. The primary asset produced from the project's CI pipeline is a container
   image for the Threeport controller that understands the needs of the custom
   workload.
0. The custom controller is deployed with the rest of the Threeport control
   plane.  Threeport is now extended to include intelligent management of
   instances of the custom workload.
0. The developers now use a custom object that requires much less configuration
   as the details of implementation are built into the custom Threeport
   controller.
0. When the custom workload is created, Threeport deploys all the components of
   the application, stitching all the components together to produce a running,

>    available application as soon as all resources are provisioned.  In this

Reviewer: ```suggestion
   available application as soon as all resources are provisioned.  In the above
```
